aten/src/ATen/native/cuda/Assert.cuh

```diff
@@ -0,0 +1,203 @@
+    T first,
+    Args... args) {
+  dst = copy_args(dst, end, first);
+  return copy_args(dst, end, args...);
+}
+
+template <typename... Args>
+C10_HOST_DEVICE __noinline__ void assert_fail(
+    CUDAAssert* assert_state,
+    int32_t error_code,
+    bool persistent,
+    uint32_t line,
+    AssertKind kind,
+    Args... args) {
+  assert(error_code != 0);
+#ifdef __CUDA_ARCH__
+  if (atomicCAS(const_cast<int32_t*>(&assert_state->error), 0, error_code) !=
+      0) {
+    return;
+  }
+#else
+  assert(false); // should never be called
+#endif
+
+  char* buffer = assert_state->buffer;
+  char* dst = buffer;
+  char* const end = dst + sizeof(assert_state->buffer);
+
+  dst = copy_args(dst, end, line, args...);
+
+  assert_state->kind = kind;
+  assert_state->line = line;
+  assert_state->length = dst ? dst - buffer : 0;
+  assert_state->persistent = persistent;
+}
+
+// handle case without format string, e.g. C10_KERNEL_ASSERT(false)
+static inline C10_HOST_DEVICE void assert_fail(
+    CUDAAssert* assert_state,
+    int32_t error_code,
+    bool persistent,
+    uint32_t line,
+    AssertKind kind) {
+  assert_fail(assert_state, error_code, persistent, line, kind, 0);
+}
+
+static inline CUDAAssert* prepare_kernel_assert() {
+  auto current_stream = getCurrentCUDAStream();
+  return current_stream.assert_state();
+}
+
+} // namespace assert
+} // namespace native
+} // namespace at
+
+#define C10_PREPARE_KERNEL_ASSERT \
+  auto __c10_assert_state = at::native::assert::prepare_kernel_assert();
+
+#define C10_KERNEL_ASSERT(code, exp, ...)  \
+  do {                                     \
+    if (!(exp)) {                          \
+      at::native::assert::assert_fail(     \
+          __c10_assert_state,              \
+          static_cast<int32_t>(code),      \
+          true, /*persistent*/             \
+          static_cast<uint32_t>(__LINE__), \
+          ##__VA_ARGS__);                  \
+      assert(exp);                         \
+      assert(false);                       \
+    }                                      \
+  } while (false)
+
+#define C10_KERNEL_ASSERT_RETURN_TYPE_(rval, code, exp, type, ...) \
+  if (!(exp)) {                                                    \
+    at::native::assert::assert_fail(                               \
+        __c10_assert_state,                                        \
+        static_cast<int32_t>(code),                                \
+        false, /*persistent*/                                      \
+        static_cast<uint32_t>(__LINE__),                           \
+        type,                                                      \
+        ##__VA_ARGS__);                                            \
+    return rval;                                                   \
+  }
+
+#define C10_KERNEL_ASSERT_RETURN_(rval, code, exp, ...) \
+  C10_KERNEL_ASSERT_RETURN_TYPE_(                       \
+      rval, code, exp, c10::cuda::AssertKind::DEFAULT, ##__VA_ARGS__)
```

Comment:
Probably best to just force users to always provide an AssertKind :)

Reply:
Yes, I created the simpler macros later because I wanted to keep the assert-invocations in the .cu files as short as possible. It would be possible to let the user pass assert-kind and and also the assert-state pointer and not used the fixed `__c10_assert_state` name (which probably is a bit ugly). I tried this before and it made the assert-invocations look quite complicated and long, e.g. 

`C10_KERNEL_ASSERT_RETURN_(ret_val, condition, error_state, file_id, assert_kind, format_id, args...);`